pkg/devspace/build/registry/local_registry.go

```diff
@@ -0,0 +1,216 @@
+package registry
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+	devspacecontext "github.com/loft-sh/devspace/pkg/devspace/context"
+	"github.com/loft-sh/devspace/pkg/devspace/kubectl"
+	"github.com/loft-sh/devspace/pkg/devspace/services/targetselector"
+	"github.com/loft-sh/devspace/pkg/util/log"
+	"github.com/mgutz/ansi"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+const (
+	LastAppliedConfigurationAnnotation = "devspace.sh/last-applied-configuration"
+)
+
+type LocalRegistry struct {
+	options     Options
+	host        string
+	servicePort *corev1.ServicePort
+}
+
+func NewLocalRegistry(options Options) *LocalRegistry {
+	return &LocalRegistry{
+		options: options,
+	}
+}
+
+func (r *LocalRegistry) IsStarted() bool {
+	return r.servicePort != nil
+}
+
+func (r *LocalRegistry) Start(ctx devspacecontext.Context) error {
+	ctx.Log().Info("Starting Local Image Registry")
+
+	if err := r.ensureNamespace(ctx); err != nil {
+		return errors.Wrap(err, "ensure namespace")
+	}
+
+	if r.options.StorageEnabled {
+		if err := r.ensureStatefulset(ctx); err != nil {
+			return errors.Wrap(err, "ensure statefulset")
+		}
+	} else {
+		if err := r.ensureDeployment(ctx); err != nil {
+			return errors.Wrap(err, "ensure deployment")
+		}
+	}
+
+	if err := r.ensureService(ctx); err != nil {
+		return errors.Wrap(err, "ensure service")
+	}
+
+	// Wait for service to have a node port
+	var err error
+	r.servicePort, err = r.waitForNodePort(ctx)
+	if err != nil {
+		return errors.Wrap(err, "wait for node port")
+	}
+
+	// Save registry host for rewriting images
+	r.host = fmt.Sprintf("localhost:%d", r.servicePort.NodePort)
+
+	// Start port forwarding
+	tctx, t := ctx.WithNewTomb()
+	pfctx := tctx.WithContext(context.Background())
+	<-t.NotifyGo(func() error {
```

Comment:
How do we catch errors here that happen during port forwarding?

Reply:
👍